### PR TITLE
Refactor/result bind

### PR DESCRIPTION
**Pull request by MBoldyrev**
### Description of the Change

Now the bound transformation function may return one of: `Result`, `Value` or any other unwrapped value that is treated like a `Value`. Also, a rvalue-reference binding specialization is added. Two functions to convert a `Result`'s `Error` or `Value` to `boost::optional` are added.

### Benefits

Hopefully allows to write a simpler code.

### Possible Drawbacks 

More code to maintain.

### CODE REVIEW

File: libs/common/result.hpp [hunk 1]
```diff
@@ -187,36 +206,119 @@ namespace iroha {
...
 
+    template <typename T>
+    constexpr bool isResult = std::is_base_of<ResultBase, T>::value;
+    template <typename T>
+    constexpr bool isValue = std::is_base_of<ValueBase, T>::value;
+    template <typename T>
+    constexpr bool isError = std::is_base_of<ErrorBase, T>::value;
+
+    /**
+     * A struct that provides the result type conversion for bind operator.
```
Comment (igor-egorov) [inline libs/common/result.hpp]: ?
```suggestion
     * A struct that provides the result of type conversion for bind operator.
```

File: libs/common/result.hpp [hunk 2]
```diff
@@ -187,36 +206,119 @@ namespace iroha {
...
+     * @tparam Transformed The type returned by value transformation function.
+     * @tparam ErrorType The type of former result's error
+     * The struct provides Result type, which is a combination of transformation
+     * function outcome and former Result error type, and a method to convert
+     * transformation function outcome to this type.
+     */
+    template <typename Transformed, typename ErrorType, typename = void>
```
Comment (igor-egorov) [inline libs/common/result.hpp]: `typename = void` - is it needed?

Comment (MBoldyrev) [reply libs/common/result.hpp]: it is needed for partial specialisation (please see below)

**Comment by MBoldyrev** [comment]
> I am not sure if we do really need that change.

It is a standalone part of changes in #84, which was taken into a separate PR (this one) to alleviate that change that became rather big.

**Pull request merged**